### PR TITLE
7zip: update to 24.07

### PR DESCRIPTION
**Pull request by aeiouaeiouaeiouaeiouaeiouaeiou**
* simplify compiler blacklist

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6.8 10K549 x86_64
Xcode 3.2.6 10M2518

macOS 10.13.6 17G14042 x86_64
Xcode 10.1 10B61

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by pmetzger** [comment]
Why the change to the flags?

**Comment by aeiouaeiouaeiouaeiouaeiouaeiou** [comment]
@pmetzger it's better to use MacPorts CFLAGS, isn't it?
FreeBSD ports use a [familiar behavior](https://codeberg.org/FreeBSD/freebsd-ports/src/commit/0002e2e8a62d0dc192b3730eaba777920d06b697/archivers/7-zip/files/patch-CPP_7zip_7zip__gcc.mak), but in my case it results in compilation without optimization flags.

File: archivers/7zip/Portfile [hunk 1]
```diff
@@ -53,10 +53,15 @@ post-patch {
     reinplace "s|LFLAGS_ALL = $\(\LFLAGS_STRIP\)\|LFLAGS_ALL =|" \
     ${worksrcpath}/C/7zip_gcc_c.mak \
     ${worksrcpath}/CPP/7zip/7zip_gcc.mak
+    # do not force -O2 flag during compilation
+    reinplace "s|CFLAGS_BASE = $\(\MY_ARCH_2\)\ -O2|CFLAGS_BASE = $\(\MY_ARCH_2\)\ ${configure.cflags}|" \
```
Comment (ryandesign) [inline archivers/7zip/Portfile:57]: `M` and ` ` are not a special characters in regular expressions and do not need a backslash before them.

`(` and `)` are special characters in extended regular expressions but you have not used `reinplace`'s `-E` flag to enable extended regular expressions, and they are not special characters in basic regular expressions.

Within quotation marks, backslashes are consumed by Tcl before reaching the regex engine, and none of these characters are special to Tcl, so all of the backslashes are unnecessary.

`$` is a special character to Tcl and to regular expressions so it should be escaped from both engines by using a double backslash in the regular expression part and a single blackslash in the replacement part.

```suggestion
    reinplace "s|CFLAGS_BASE = \\$(MY_ARCH_2) -O2|CFLAGS_BASE = \$(MY_ARCH_2) ${configure.cflags}|" \
```

Please apply these suggestions to the existing `reinplace`s for `LDFLAGS` and `LFLAGS_ALL`.

Comment (aeiouaeiouaeiouaeiouaeiouaeiou) [reply archivers/7zip/Portfile]: `Error: Failed to patch 7zip: can't read "(MY_ARCH_2)": no such variable`

I don't think this behavior was expected.

**Comment by pmetzger** [comment]
I suppose using explicit CFLAGS is better, but the performance of 7zip is important and you need to make sure it's using -O2 regardless.

File: archivers/7zip/Portfile [hunk 2]
```diff
@@ -53,10 +53,15 @@ post-patch {
     reinplace "s|LFLAGS_ALL = $\(\LFLAGS_STRIP\)\|LFLAGS_ALL =|" \
     ${worksrcpath}/C/7zip_gcc_c.mak \
     ${worksrcpath}/CPP/7zip/7zip_gcc.mak
+    # do not force -O2 flag during compilation
+    reinplace "s|CFLAGS_BASE = $\(\MY_ARCH_2\)\ -O2|CFLAGS_BASE = $\(\MY_ARCH_2\)\ -O2 ${configure.cflags}|" \
```
Comment (pmetzger) [inline archivers/7zip/Portfile:57]: I don't get why there are backslashes in front of the Ms and the spaces.

Comment (aeiouaeiouaeiouaeiouaeiouaeiou) [reply archivers/7zip/Portfile:57]: As I wrote above, without those backslashes it gives a "no such variable" error (both locally and CI).

Comment (pmetzger) [reply archivers/7zip/Portfile:57]: Please figure out why. "M" is not a special character etc. Something weird is happening and mysteries are bad.

**Pull request merged**